src/backend/utils/cache/lsyscache.c

```diff
@@ -3051,6 +3051,16 @@ get_typcollation(Oid typid)
 		Oid			result;
 
 		result = typtup->typcollation;
+		if (IsNormalProcessingMode() && OidIsValid(typtup->typcollation) &&
+			sql_dialect == SQL_DIALECT_TSQL && pg_strcasecmp(get_namespace_name(typtup->typnamespace), "sys") == 0)
+		{
+			/*
+			 * Always set CLUSTER_COLLATION_OID() for babelfish collatable types so that
+			 * we can set collation according to database or server level later.
+			 */
+			result = CLUSTER_COLLATION_OID();
+		}
+
```

Comment:
Let's just call typeTypeCollation by default.
1. Why do we need IsNormalProcessingMode() check here?
2. There're multiple callers of get_typcollation where we're again setting collation to CLUSTER_COLLATION_OID. Let's remove those codes that are no longer needed because of this change.
3. How does it affect inter-operability? I'm concerned that we're using different collations depending on the dialect. Won't that output wrong data in PG dialect?

Reply:
Also, what are the scenarios when a built-in datatype is created with non-default collation? The built-in datatypes are in our control. So, why not fix the collation when we create the same!!

Reply:
> Let's just call typeTypeCollation by default.

Instead I am calling hook defined to handle default collation as there is special handling required for string Const node.

> Why do we need IsNormalProcessingMode() check here?

This part of code is also getting called during bootstrap so this check was added to safely handle that case. But since we are moving the handling to extension through hook, this check is no more required and hence removed it. 

> There're multiple callers of get_typcollation where we're again setting collation to CLUSTER_COLLATION_OID. 

Let's remove those codes that are no longer needed because of this change.
Yes, code cleanup is done with additional handling for some cases, ie., handling default collation of user defined data type. Default collation of UDF should always be same as database collation. check https://github.com/babelfish-for-postgresql/babelfish_extensions/pull/2739/commits/d0417a36e5917c2edc750531f7dcb7d0b7198665 and https://github.com/babelfish-for-postgresql/babelfish_extensions/pull/2739/commits/c50e653e127eb90a11505da69b0419093cdb3562


>How does it affect inter-operability? I'm concerned that we're using different collations depending on the dialect. Won't that output wrong data in PG dialect?

There will not be any impact on inter-operability as collation of collatable data is always defined. So even in PG dialect, appropriate collation will be picked up.

> Also, what are the scenarios when a built-in datatype is created with non-default collation? The built-in datatypes are in our control. So, why not fix the collation when we create the same!!

built-in datatypes are always defined with collation same as server_collation_name. But with database level collation being supported (which might be different from server_collation_name), we need this handling to assign database level collation whenever object collation is not explicitly stated.